### ISSUE TITLE
Speed improvement suggestion: array resizing

### ISSUE THREAD

**Issue by bkDJ** [uiua-lang/uiua#450]
Example:

```uiua
◌ ⍥(⊂⚂)1e5 []
```

Obviously that is not how you would write code for the same result. But sometimes, a `join` in a `do`/`repeat` is the simplest way to accomplish something (such as generating all permutations of a list that satisfy a certain condition). When I run the above with `uiua run 'file.ua' --time-instrs | grep ⊂`, the time taken seems to increase proportionally to the current length of the array on every single call.

```sh
# first few lines
  ⏲    0.01ms - ⊂
  ⏲    0.00ms - ⊂
  ⏲    0.00ms - ⊂
# last few lines
  ⏲    0.36ms - ⊂
  ⏲    0.36ms - ⊂
  ⏲    0.40ms - ⊂
```

How do you feel about having the backing Rust array pre-allocate higher amounts (e.g. instead of "exact fit", leave some head-room up to the next power of two? Ruby does [something similar](https://ivoanjo.me/blog/2021/02/11/looking-into-array-memory-usage/).) when an array is too small to fit its next element?

**Comment by kaikalii** [comment]
I agree this should be fixed.

If you `join` to the end instead of the front, does the extra time go away?

**Comment by kaikalii** [comment]
> If you `join` to the end instead of the front, does the extra time go away?

I did some testing and yes, `join`ing to the end has no extra cost.

I'm going to push a change that fixes `join`ing to the front.
It will still technically be O(n), since all the existing elements have to get shifted over. But in my testing, it's 1 or 2 orders of magnitude faster since there is no longer an allocator call on ever `join`.

**Comment by kaikalii** [comment]
Fixed in 67634f73ca02bd5470.
This fix only applies to `join`, but there may be opportunities for other similar optimizations.

**Comment by bkDJ** [comment]
Super cool, thanks!! And I'll take care to `join` `flip` when in a loop.

BTW, if there is still a known performance cost for prepending instead of appending, I'd also recommend the docs call that out.

Thanks again for the lightning fix.
